cmake/BLTMacros.cmake

```diff
@@ -258,36 +258,32 @@ macro(blt_register_library)
     mark_as_advanced(BLT_${uppercase_name}_TREAT_INCLUDES_AS_SYSTEM)
 
     if( ENABLE_FORTRAN AND arg_FORTRAN_MODULES )
-        set(BLT_${uppercase_name}_FORTRAN_MODULES ${arg_INCLUDES} CACHE PATH "" FORCE)
+        set(BLT_${uppercase_name}_FORTRAN_MODULES ${arg_INCLUDES} CACHE LIST "" FORCE)
         mark_as_advanced(BLT_${uppercase_name}_FORTRAN_MODULES)
     endif()
 
     if( arg_LIBRARIES )
-        set(BLT_${uppercase_name}_LIBRARIES ${arg_LIBRARIES} CACHE PATH "" FORCE)
+        set(BLT_${uppercase_name}_LIBRARIES ${arg_LIBRARIES} CACHE LIST "" FORCE)
     else()
-        set(BLT_${uppercase_name}_LIBRARIES "BLT_NO_LIBRARIES" CACHE PATH "" FORCE)
+        # This stops cmake from falling back on adding -l<library name>
+        # to the command line for BLT registered libraries which are not
+        # actual CMake targets
+        set(BLT_${uppercase_name}_LIBRARIES "BLT_NO_LIBRARIES" CACHE STRING "" FORCE)
     endif()
-
     mark_as_advanced(BLT_${uppercase_name}_LIBRARIES)
 
     if( arg_COMPILE_FLAGS )
-        set(BLT_${uppercase_name}_COMPILE_FLAGS ${arg_COMPILE_FLAGS} CACHE PATH "" FORCE)
-    else()
-        set(BLT_${uppercase_name}_COMPILE_FLAGS "BLT_NO_COMPILE_FLAGS" CACHE PATH "" FORCE)
+        set(BLT_${uppercase_name}_COMPILE_FLAGS "${arg_COMPILE_FLAGS}" CACHE STRING "" FORCE)
+        mark_as_advanced(BLT_${uppercase_name}_COMPILE_FLAGS)
     endif()
 
-    mark_as_advanced(BLT_${uppercase_name}_COMPILE_FLAGS)
-
     if( arg_LINK_FLAGS )
-        set(BLT_${uppercase_name}_LINK_FLAGS ${arg_LINK_FLAGS} CACHE PATH "" FORCE)
-    else()
-        set(BLT_${uppercase_name}_LINK_FLAGS "BLT_NO_LINK_FLAGS" CACHE PATH "" FORCE)
+        set(BLT_${uppercase_name}_LINK_FLAGS "${arg_LINK_FLAGS}" CACHE STRING "" FORCE)
+        mark_as_advanced(BLT_${uppercase_name}_LINK_FLAGS)
     endif()
 
-    mark_as_advanced(BLT_${uppercase_name}_LINK_FLAGS)
-
     if( arg_DEFINES )
-        set(BLT_${uppercase_name}_DEFINES ${arg_DEFINES} CACHE PATH "" FORCE)
+        set(BLT_${uppercase_name}_DEFINES ${arg_DEFINES} CACHE LIST "" FORCE)
```

Comment:
Should this one be a ``STRING``? Or is cmake able to convert it to a space-separated string when it adds it to the compiler command?

Reply:
It's a list.  They shouldnt be given as string of -DHAVE_FOO -DHAVE_BAR options... they should be given as a list of HAVE_FOO HAVE_BAR